### DISCUSSION TITLE
On `ModuleDecl`

### DISCUSSION THREAD

**Discussion by Kyuuhachi** [swc-project/swc#8753] [General]
I'm confused/annoyed at the design of `ModuleDecl`, which seems to be largely imported from estree. Ignoring things like spans and TypeScript, it basically boils down to:

```rs
enum ModuleDecl {
	Import {
		specifiers: Vec<enum {
			Named { local: Ident, imported: Option<IdentOrStr> },
			Default { local: Ident },
			Namespace { local: Ident },
		}>,
		src: Str,
	},
	ExportDecl {
		decl: Decl,
	},
	ExportNamed {
		specifiers: Vec<enum {
			Namespace { name: IdentOrStr },
			Default { exported: Ident },
			Named { orig: IdentOrStr, exported: Option<IdentOrStr> },
		}>,
		src: Option<Str>,
	},
	ExportDefaultDecl {
		decl: enum {
			Class(ClassExpr),
			Fn(FnExpr),
			TsInterfaceDecl(TsInterfaceDecl),
		},
	},
	ExportDefaultExpr {
		expr: Expr,
	},
	ExportAll {
		src: Str,
	}
}
```

There are several points I'm confused about:

- Why are exports and reexports (`export {};` and `export {} from "...";`) joined? They have different both syntax and semantics. I imagine it makes parsing easier, though. If anything, `import {} from "...";` and `export {} from "...";` are are quite similar.
- Why are `ExportDefaultDecl` and `ExportDefaultExpr` split? I don't understand why `export default function () {};` wouldn't be a `ExportDefaultExpr`. Only complicating factor seems to be `TsInterfaceDecl`, which I am not knowledgeable enough to speak about.
- Most importantly, why the `Vec<Specifier>` scheme, rather than just fixed fields? It makes any processing require awkward loops.

While I acknowledge that I may be oversimplifying certain parts, in my mind, the the ideal design would be something closer to

```rs
enum ModuleDecl {
	Import {
		default: Option<Ident>,
		namespace: Option<enum {
			All { local: Ident },
			Named {
				items: Vec<struct {
					imported: Option<IdentOrStr>,
					local: Ident,
				}>,
			},
		}>,
		src: Str,
	},

	ExportNamed {
		namespace: Vec<struct {
			local: Ident,
			exported: Option<IdentOrStr>,
		}>,
	},
	ExportDecl {
		decl: Decl,
	},
	ExportDefault {
		expr: Expr,
	},

	ReexportNamed {
		namespace: enum {
			All { local: Ident },
			Named {
				items: Vec<struct {
					imported: Option<IdentOrStr>,
					exported: IdentOrStr,
				}>,
			},
		},
		src: Str,
	},
	ReexportAll {
		src: Str,
	},
}
```

I believe this structure more closely reflects the actual set of allowed productions, making it less error-prone; and removing the `Vec<Specifier>` scheme makes it significantly easier to work with. The obvious downside is that it's a majorly breaking change.

Should you judge this to be a worthwhile change, I would be willing to attempt to implement it.

**Comment by kdy1** [comment]
1. It's AST. Those two are syntactically very similar.
2. Export Default Decl and Export Default Expr has different hoistng semantics.
3. It's based on estree, and your AST looks more complex than the current one.
